tianshou/data/buffer/her.py

```diff
@@ -0,0 +1,180 @@
+from typing import Any, Callable, List, Optional, Tuple, Union
+
+import numpy as np
+
+from tianshou.data import Batch, ReplayBuffer
+
+
+class HERReplayBuffer(ReplayBuffer):
+    """Implementation of Hindsight Experience Replay. arXiv:1707.01495.
+
+    Currently support only 'future' strategy of HER.
+
+    :param int size: the size of the replay buffer.
+    :param compute_reward_fn: a function that takes 3 arguments: \
+        `acheived_goal`, `desired_goal`, `info` and returns the reward(s).
+        Note that the goal arguments can have extra batch_size dimension and in that \
+        case, the rewards of size batch_size should be returned
+    :param int horizon: the maximum number of steps in an episode.
+    :param int future_k: the 'k' parameter introduced in the paper. In short, there \
+        will be at most k episodes that are re-written for every 1 unaltered episode \
+        during the sampling.
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.data.ReplayBuffer` for other APIs' usage.
+    """
+
+    def __init__(
+        self,
+        size: int,
+        compute_reward_fn: Callable[[np.ndarray, np.ndarray, dict], np.ndarray],
+        horizon: int,
+        future_k: float = 8.0,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(size, **kwargs)
+        self.horizon = horizon
+        self.future_p = 1 - 1 / future_k
+        self.compute_reward_fn = compute_reward_fn
+        self._original_meta = Batch()
+        self._altered_indices = np.array([])
+
+    def _restore_cache(self) -> None:
+        """Write cached original meta back to `self._meta`.
+
+        It's called everytime before 'writing', 'sampling' or 'saving' the buffer.
+        """
+        if not hasattr(self, '_altered_indices'):
+            return
+
+        if self._altered_indices.size == 0:
+            return
+        self._meta[self._altered_indices] = self._original_meta
+        # Clean
+        del self._original_meta, self._altered_indices
```

Comment:
no need to manually delete them because it has been overwritten in the next line?

Reply:
fixed: https://github.com/thu-ml/tianshou/pull/753/commits/63e65c087fa4c070861f12ba4d5fea130fc56740